File: hamiltonian.m
%Hamiltonian - The Hamiltonian described in [1] solved with SeDuMi
%
%[1] Corboz, P.; Evenbly, G.; Verstraete, F. & Vidal, G. (2009), 
%    Simulation of interacting fermions with entanglement renormalization.
%    arXiv:0904.4151
%
% Other m-files required: Yalmip, SeDuMi, ncmoments.m, momentstosdpa.m
%                         getneighbors.m lineartolattice.m
% MAT-files required: none
%
% Author: Peter Wittek
% April 2013; Last revision: 11-Apr-2013


%------------- BEGIN CODE --------------

% Explicit definition that i will be sqrt(-1) in the code below
i=sqrt(-1);

% The dimension of the lattice in each direction
ld=2;

% Define non-commutative variables.
% Twice the number is required, as imaginary components
% will be handled manually. An operator C will be represented as
% C=X(r)+i*X(r+1)
n = 2*ld*ld;
X = ncvar(n,1);

% Parameters for the Hamiltonian
gamma=1;lambda=2;

obj=0; equalities = [];
% Constructing the Hamiltonian and the constraints
for r=1:2:n
    % Neighbors of r in the lattice
    neighbors=getneighbors(r, ld);
    for k=1:size(neighbors)
        s=neighbors(k);
        % The Hamiltonian
        obj = obj + ...
          (X(r)-i*X(r+1))*(X(s)+i*X(s+1))+(X(s)-i*X(s+1))*(X(r)+i*X(r+1)) + ...
          gamma*((X(r)-i*X(r+1))*(X(s)-i*X(s+1))+(X(s)+i*X(s+1))*(X(r)+i*X(r+1))) + ...
          2*lambda*((X(r)-i*X(r+1))*(X(r)+i*X(r+1)));
        % Equality constraints
        equalities = [ equalities; ...
          (X(r)+i*X(r+1))*(X(s)+i*X(s+1))+(X(s)+i*X(s+1))*(X(r)+i*X(r+1));...
          (X(r)+i*X(r+1))*(X(s)-i*X(s+1))+(X(s)-i*X(s+1))*(X(r)+i*X(r+1))];

    end
end

% There are no inequality constraints
inequalities = [];

%Order of relaxation (must be at least 2)
order = 2;

[Fmoments, M] = ncmoments(X, inequalities, equalities, order);

yalmip_options = sdpsettings('solver','sedumi','removeequalities ', 1,'relax',1,'saveduals',0);
sol=solvesdp(Fmoments,obj,yalmip_options);

% Obtaining the moment matrix
Mrel = relaxdouble(M{order+1});

File: hamiltonian.py
# -*- coding: utf-8 -*-
"""
Exporting a Hamiltonian ground state problem to SDPA. The Hamiltonian 
is described in the following paper:

Corboz, P.; Evenbly, G.; Verstraete, F. & Vidal, G. (2009), 
Simulation of interacting fermions with entanglement renormalization.
arXiv:0904.4151

Created on Fri May 10 09:45:11 2013

@author: Peter Wittek
"""
import time
from sympy.core import S
from sympy.physics.quantum.operator import HermitianOperator
from sympy.physics.quantum.dagger import Dagger
from sdprelaxation import get_relaxation
from ncutils import get_neighbors

#Lattice dimension
lattice_dimension = 2
n_vars = lattice_dimension * lattice_dimension
#Order of relaxation
order = 2

# Parameters for the Hamiltonian
gam, lam = 1, 2

#Get Hermitian variables
C = [0]*n_vars
for i in xrange(n_vars):
    C[i] = HermitianOperator('C%s' % i)

hamiltonian = 0
for r in xrange(n_vars):
    hamiltonian -= 2*lam*Dagger(C[r])*C[r]
    neighbors = get_neighbors(r, lattice_dimension)
    for s in neighbors:
        hamiltonian += Dagger(C[r])*C[s]+Dagger(C[s])*C[r]
        hamiltonian -= gam*(Dagger(C[r])*Dagger(C[s])+C[s]*C[r])

monomial_substitution = {}
equalities = []
for r in xrange(n_vars):
    for s in xrange(n_vars):
        if not r == s:
            monomial_substitution[C[r] * C[s]] = -C[s] * C[r]
        else:
            equalities.append(C[r]*Dagger(C[s]) + Dagger(C[s])*C[r] - S.One)

inequalities = []

time0 = time.time()
#Obtain SDP relaxation
print "Obtaining SDP relaxation..."
prob = get_relaxation(C, hamiltonian, inequalities, equalities, 
                      monomial_substitution, order)
#Export relaxation to SDPA format
print "Writing to disk..."
prob.write_to_file('hamiltonian.dat-s')

print '%d %0.2f s' % (lattice_dimension, (time.time()-time0))
